README
conky(1)                                                              conky(1)



NAME
       conky - A system monitor for X originally based on the torsmo code, but
       more kickass. It just keeps on given'er. Yeah.

SYNOPSIS
       conky [ options ]

DESCRIPTION
       Conky is a system monitor for X originally based on torsmo.  Since  its
       inception,  Conky has changed significantly from its predecessor, while
       maintaining simplicity and  configurability.  Conky  can  display  just
       about  anything,  either on your root desktop or in its own window. Not
       only does Conky have many built-in objects, it can  also  display  just
       about any piece of information by using scripts and other external pro-
       grams.

       Conky has more than 250 built  in  objects,  including  support  for  a
       plethora  of OS stats (uname, uptime, CPU usage, mem usage, disk usage,
       "top" like process stats, and network monitoring, just to name a  few),
       built in IMAP and POP3 support, built in support for many popular music
       players (MPD, XMMS2, BMPx, Audacious), and much much more.   Conky  can
       display  this  info  either  as text, or using simple progress bars and
       graph widgets, with different fonts and colours.

       We are always looking for help, whether  its  reporting  bugs,  writing
       patches,  or  writing docs. Please use the facilities at SourceForge to
       make bug reports, feature requests, and  submit  patches,  or  stop  by
       #conky on irc.freenode.net if you have questions or want to contribute.

       Thanks for your interest in Conky.

COMPILING
       For  users compiling from source on a binary distro, make sure you have
       the X development libraries installed  (Unless  you  provide  configure
       with  "--disable-x11").  This  should  be  a package along the lines of
       "libx11-dev" or "xorg-x11-dev" for X11 libs, and similar "-dev"  format
       for  the other libs required (depending on your configure options). You
       should be able to see which extra packages you need to install by read-
       ing  errors  that  you  get  from './configure'. You can enable/disable
       stuff by giving options to configure, but be  careful  with  disabling.
       For  example:  with --disable-math you won't get errors but logarithmic
       graphs will be normal graphs and gauges will miss their line.

       Conky has (for some time) been available in the  repositories  of  most
       popular  distributions.  Here  are some installation instructions for a
       few:

       Gentoo users -- Conky is in Gentoo's  Portage...   simply  use  "emerge
       app-admin/conky" for installation.

       Debian,  etc. users -- Conky should be in your repositories, and can be
       installed by doing "aptitude install conky".

       Example to compile and run Conky with  all  optional  components  (note
       that some configure options may differ for your system):

       sh autogen.sh # Only required if building from the git repo

       ./configure       --prefix=/usr      --mandir=/usr/share/man      --in-
       fodir=/usr/share/info --datadir=/usr/share  --sysconfdir=/etc  --local-
       statedir=/var/lib --disable-own-window --enable-audacious[=yes|no|lega-
       cy] --enable-bmpx --disable-hddtemp --disable-mpd --enable-xmms2 --dis-
       able-portmon   --disable-network  --enable-debug  --disable-x11  --dis-
       able-double-buffer --disable-xdamage --disable-xft

       make

       make install # Optional

       src/conky

       Conky has been tested to be compatible with C99 C, however it  has  not
       been tested with anything other than gcc, and is not guaranteed to work
       with other compilers.

YOU SHOULD KNOW
       Conky is generally very good on resources. That said, the more you  try
       to make Conky do, the more resources it is going to consume.

       An easy way to force Conky to reload your ~/.conkyrc: "killall -SIGUSR1
       conky". Saves you the trouble of having to kill and then  restart.  You
       can now also do the same with SIGHUP.

OPTIONS
       Command  line  options override configurations defined in configuration
       file.

       -v | -V | --version
              Prints version and exits


       -q | --quiet
              Run Conky in 'quiet mode' (ie. no output)


       -D | --debug
              Increase debugging output, ie. -DD for more debugging


       -a | --alignment= ALIGNMENT
              Text alignment on  screen,  {top,bottom,middle}_{left,right,mid-
              dle}  or none. Can also be abbreviated with first chars of posi-
              tion, ie. tr for top_right.


       -b | --double-buffer
              Use double buffering (eliminates "flicker")


       -c | --config= FILE
              Config file to load instead of $HOME/.conkyrc


       -C | --print-config
              Print builtin default config to stdout. See also the section EX-
              AMPLES for more information.


       -d | --daemonize
              Daemonize Conky, aka fork to background


       -f | --font= FONT
              Font to use


       -h | --help
              Prints command line help and exits


       -o | --own-window
              Create own window to draw


       -t | --text= TEXT
              Text to render, remember single quotes, like -t ' $uptime '


       -p | --pause= SECONDS
              Time to pause before actually starting Conky


       -u | --interval= SECONDS
              Update interval


       -w | --window-id= WIN_ID
              Window id to draw


       -X | --display= DISPLAY
              X11 display to use


       -x X_COORDINATE
              X position


       -y Y_COORDINATE
              Y position


       -i COUNT
              Number of times to update Conky (and quit)


CONFIGURATION SETTINGS
       Default    configuration    file    location   is   $HOME/.conkyrc   or
       ${sysconfdir}/conky/conky.conf. On most systems,  sysconfdir  is  /etc,
       and you can find the sample config file there (/etc/conky/conky.conf).

       You  might  want  to copy it to $HOME/.conkyrc and then start modifying
       it. Other configs can be found at http://conky.sf.net/

       TEXT   After this begins text to be formatted on screen.  Backslash (\)
              escapes  newlines  in  the  text section. This can be useful for
              cleaning up config files where conky is used to  pipe  input  to
              dzen2.


       alignment
              Aligned position on screen, may be top_left, top_right, top_mid-
              dle, bottom_left, bottom_right, bottom_middle, middle_left, mid-
              dle_middle, middle_right, or none (also can be abbreviated as tl,
              tr, tm, bl, br, bm, ml, mm, mr). See also gap_x and gap_y.


       append_file
              Append the file given as argument.


       background
              Boolean value, if true, Conky will be forked to background  when
              started.


       border_inner_margin
              Inner border margin in pixels (the margin between the border and
              text).


       border_outer_margin
              Outer border margin in pixels (the margin between the border and
              the edge of the window).


       border_width
              Border width in pixels.


       colorN Predefine a color for use inside TEXT segments.  Substitute N by
              a digit between 0 and 9, inclusively. When specifying the  color
              value in hex, omit the leading hash (#).


       cpu_avg_samples
              The number of samples to average for CPU monitoring.


       default_bar_size
              Specify  a  default  width  and  height for bars.  Example: 'de-
              fault_bar_size 0 6'. This is particularly useful for execbar and
              execibar as they do not take size arguments.


       default_color
              Default color and border color


       default_gauge_size
              Specify  a  default  width and height for gauges.  Example: 'de-
              fault_gauge_size 25 25'. This is particularly useful  for  exec-
              gauge and execigauge as they do not take size arguments


       default_graph_size
              Specify  a  default  width and height for graphs.  Example: 'de-
              fault_graph_size 0 25'. This is particularly  useful  for  exec-
              graph and execigraph as they do not take size arguments


       default_outline_color
              Default outline color


       default_shade_color
              Default shading color and border's shading color


       disable_auto_reload
              Enable to disable the inotify-based auto config reload feature.


       diskio_avg_samples
              The number of samples to average for disk I/O monitoring.


       display
              Specify an X display to connect to.


       double_buffer
              Use the Xdbe extension? (eliminates flicker) It is highly recom-
              mended to use own window with this one so double buffer won't be
              so big.


       draw_borders
              Draw borders around text?


       draw_graph_borders
              Draw borders around graphs?


       draw_outline
              Draw outlines?


       draw_shades
              Draw shades?


       extra_newline
              Put  an  extra newline at the end when writing to stdout, useful
              for writing to awesome's wiboxes.


       font   Font name in X, xfontsel can be used to get a nice font


       format_human_readable
              If enabled, values which are in bytes will be printed  in  human
              readable  format  (i.e., KiB, MiB, etc). If disabled, the number
              of bytes is printed instead.


       gap_x  Gap, in pixels, between right or left border of screen, same  as
              passing  -x  at command line, e.g. gap_x 10.  For other position
              related stuff, see 'alignment'.


       gap_y  Gap, in pixels, between top or bottom border of screen, same  as
              passing  -y  at command line, e.g. gap_y 10.  For other position
              related stuff, see 'alignment'.


       hddtemp_host
              Hostname  to  connect  to  for  hddtemp  objects.  Defaults   to
              "127.0.0.1".


       hddtemp_port
              Port to use for hddtemp connections. Defaults to 7634.


       if_up_strictness
              How  strict  should if_up be when testing an interface for being
              up? The value is one of up, link or address, to  check  for  the
              interface being solely up, being up and having link or being up,
              having link and an assigned IP address.


       imap   Default global IMAP server. Arguments are: "host user  pass  [-i
              interval  (in  seconds)]  [-f 'folder'] [-p port] [-e 'command']
              [-r retries]". Default port is 143, default folder  is  'INBOX',
              default interval is 5 minutes, and default number of retries be-
              fore giving up is 5. If the password is  supplied  as  '*',  you
              will be prompted to enter the password when Conky starts.


       imlib_cache_flush_interval
              Interval (in seconds) to flush Imlib2 cache.


       imlib_cache_size
              Imlib2  image  cache  size, in bytes. Defaults to 4MiB. Increase
              this value if you use $image lots. Set to 0 to disable the image
              cache.


       lua_draw_hook_post function_name [function arguments]
              This  function, if defined, will be called by Conky through each
              iteration after drawing to the window. Requires X support. Takes
              any  number  of  optional  arguments.  Use this hook for drawing
              things on top of what Conky draws. Conky puts 'conky_' in  front
              of  function_name to prevent accidental calls to the wrong func-
              tion unless you place 'conky_' in front of it yourself.


       lua_draw_hook_pre function_name [function arguments]
              This function, if defined, will be called by Conky through  each
              iteration  before  drawing  to  the  window. Requires X support.
              Takes any number of optional arguments. Use this hook for  draw-
              ing  things  on  top of what Conky draws. Conky puts 'conky_' in
              front of function_name to prevent accidental calls to the  wrong
              function unless you place 'conky_' in front of it yourself.


       lua_load
              Loads the Lua scripts separated by spaces.


       lua_shutdown_hook function_name [function arguments]
              This  function,  if defined, will be called by Conky at shutdown
              or when the configuration is reloaded. Use this hook to clean up
              after  yourself, such as freeing memory which has been allocated
              by external libraries via Lua. Conky puts 'conky_' in  front  of
              function_name  to prevent accidental calls to the wrong function
              unless you place 'conky_' in front of it yourself.


       lua_startup_hook function_name [function arguments]
              This function, if defined, will be called by Conky at startup or
              when the configuration is reloaded.  Use this hook to initialize
              values, or for any run-once applications. Conky puts 'conky_' in
              front  of function_name to prevent accidental calls to the wrong
              function unless you place 'conky_' in front of it yourself.


       mail_spool
              Mail spool for mail checking


       max_port_monitor_connections
              Allow each port monitor to track at most this  many  connections
              (if 0 or not set, default is 256)


       max_specials
              Maximum  number  of special things, e.g. fonts, offsets, aligns,
              etc. (default is 512)


       max_text_width width
              When a line in the output contains 'width'  chars  and  the  end
              isn't  reached,  the  next char will start on a new line. If you
              want to make sure that lines don't get broken, set 'width' to 0


       max_user_text bytes
              Maximum size of user text buffer, i.e. layout below TEXT line in
              config file (default is 16384 bytes)


       maximum_width pixels
              Maximum width of window


       minimum_size width (height)
              Minimum size of window


       mpd_host
              Host of MPD server


       mpd_password
              MPD server password


       mpd_port
              Port of MPD server


       music_player_interval
              Music  player thread update interval (defaults to Conky's update
              interval)


       net_avg_samples
              The number of samples to average for net data


       no_buffers
              Subtract (file system) buffers from used memory?


       out_to_console
              Print text to stdout.


       out_to_ncurses
              Print text in the console, but use ncurses  so  that  conky  can
              print the text of a new update over the old text. (In the future
              this will provide more useful things)


       out_to_stderr
              Print text to stderr.


       out_to_x
              When set to no, there will be no output in X  (useful  when  you
              also  use things like out_to_console). If you set it to no, make
              sure that it's placed before all other X-related  setting  (take
              the  first line of your configfile to be sure). Default value is
              yes


       override_utf8_locale
              Force UTF8? requires XFT


       overwrite_file
              Overwrite the file given as argument.


       own_window
              Boolean, create own window to draw?


       own_window_class
              Manually set the WM_CLASS name. Defaults to "Conky".


       own_window_colour colour
              If own_window_transparent no, set a specified background  colour
              (defaults to black). Takes either a hex value (e.g. ffffff, note
              the lack of '#') or a valid RGB name (see /usr/lib/X11/rgb.txt)


       own_window_hints undecorated,below,above,sticky,skip_taskbar,skip_pager
              If own_window is yes, you may use these window manager hints  to
              affect  the way Conky displays. Notes: Use own_window_type desk-
              top as another way to implement many of these hints  implicitly.
              If  you  use own_window_type override, window manager hints have
              no meaning and are ignored.


       own_window_title
              Manually set the window name. Defaults to "<hostname> - conky".


       own_window_argb_visual
              Boolean, use ARGB visual? ARGB can be used for real  transparen-
              cy,  note  that  a composite manager is required for real trans-
              parency. This option will not work as desired (in most cases) in
              conjunction with 'own_window_type override'.


       own_window_argb_value
              When ARGB visuals are enabled, this use this to modify the alpha
              value used. Valid range is 0-255, where 0 is 0% opacity, and 255
              is 100% opacity. Note that if own_window_transparent is enabled,
              this value has no effect.


       own_window_transparent
              Boolean, set transparency? If ARGB visual is enabled, sets back-
              ground opacity to 0%.


       own_window_type
              if  own_window  is  yes,  you  may specify type normal, desktop,
              dock, panel or override (default: normal).  Desktop windows  are
              special windows that have no window decorations; are always vis-
              ible on your desktop; do not appear in your  pager  or  taskbar;
              and  are  sticky  across  all  workspaces. Panel windows reserve
              space along a desktop edge, just like panels and taskbars,  pre-
              venting  maximized  windows  from  overlapping them. The edge is
              chosen based on the alignment option. Override windows  are  not
              under the control of the window manager. Hints are ignored. This
              type of window can be useful for certain situations.


       pad_percents
              Pad percentages to this many decimals (0 = no padding)


       pop3   Default global POP3 server. Arguments are: "host user  pass  [-i
              interval  (in  seconds)] [-p port] [-e 'command'] [-r retries]".
              Default port is 110, default interval is 5 minutes, and  default
              number of retries before giving up is 5. If the password is sup-
              plied as '*', you will be prompted to enter  the  password  when
              Conky starts.


       short_units
              Shortens units to a single character (kiB->k, GiB->G, etc.). De-
              fault is off.


       show_graph_range
              Shows the time range covered by a graph.


       show_graph_scale
              Shows the maximum value in scaled graphs.


       stippled_borders
              Border stippling (dashing) in pixels


       temperature_unit
              Desired output unit of all objects displaying a temperature. Pa-
              rameters  are either "fahrenheit" or "celsius". The default unit
              is degree Celsius.


       templateN
              Define a template for later use inside TEXT segments. Substitute
              N  by  a  digit  between  0 and 9, inclusively. The value of the
              variable is being inserted into the stuff below TEXT at the cor-
              responding position, but before some substitutions are applied:

              '\n' -> newline
              '\\' -> backslash
              '\ ' -> space
              '\N' -> template argument N


       text_buffer_size bytes
              Size  of  the  standard text buffer (default is 256 bytes). This
              buffer is used for intermediary text, such as individual  lines,
              output  from $exec vars, and various other variables. Increasing
              the size of this buffer can drastically reduce  Conky's  perfor-
              mance,  but  will  allow for more text display per variable. The
              size of this buffer cannot be smaller than the default value  of
              256 bytes.


       times_in_seconds
              If true, variables that output times output a number that repre-
              sents seconds. This doesn't affect $time, $tztime and $utime


       top_cpu_separate
              If true, cpu in top will show usage of one processor's power. If
              false,  cpu  in top will show the usage of all processors' power
              combined.


       top_name_width
              Width for $top name value (defaults to 15 characters).


       total_run_times
              Total number of times for Conky to update before quitting.  Zero
              makes Conky run forever


       update_interval seconds
              Update interval


       update_interval_on_battery seconds
              Update interval when running on batterypower


       uppercase
              Boolean value, if true, text is rendered in upper case


       use_spacer
              Adds spaces around certain objects to stop them from moving oth-
              er things around. Arguments are left, right, and none (default).
              The   old  true/false  values  are  deprecated  and  default  to
              right/none respectively. Note that this only helps  if  you  are
              using a mono font, such as Bitstream Vera Sans Mono.


       use_xft
              Use Xft (anti-aliased font and stuff)


       xftalpha
              Alpha of Xft font. Must be a value at or between 1 and 0.


       xftfont
              Xft font to use.


OBJECTS/VARIABLES
       Colours  are parsed using XParsecolor(), there might be a list of them:
       /usr/share/X11/rgb.txt.  Colour can be also in #rrggbb format (hex).

       Some objects may create threads, and sometimes these threads  will  not
       be  destroyed  until  Conky  terminates.  There is no way to destroy or
       clean up threads while Conky is running. For example, if you use an MPD
       variable,  the  MPD  thread  will  keep  running until Conky dies. Some
       threaded objects will use one of the parameters as a 'key', so that you
       only  have  1 relevant thread running (for example, the $curl, $rss and
       $weather objects launch one thread per URI).

       acpiacadapter (adapter)
              ACPI ac adapter state. On linux, the  adapter  option  specifies
              the  subfolder  of  /sys/class/power_supply containing the state
              information (defaults to "AC"). Other systems ignore it.


       acpifan
              ACPI fan state


       acpitemp
              ACPI temperature in C.


       addr (interface)
              IP address for an interface, or "No Address" if  no  address  is
              assigned.


       addrs (interface)
              IP  addresses for an interface (if one - works like addr). Linux
              only.


       adt746xcpu
              CPU temperature from therm_adt746x


       adt746xfan
              Fan speed from therm_adt746x


       alignc (num)
              Align text to centre


       alignr (num)
              Right-justify text, with space of N


       apcupsd host port
              Sets up the connection to apcupsd daemon.  Prints  nothing,  de-
              faults to localhost:3551


       apcupsd_cable
              Prints the UPS connection type.


       apcupsd_charge
              Current battery capacity in percent.


       apcupsd_lastxfer
              Reason for last transfer from line to battery.


       apcupsd_linev
              Nominal input voltage.


       apcupsd_load
              Current load in percent.


       apcupsd_loadbar
              Bar showing current load.


       apcupsd_loadgauge (height),(width)
              Gauge that shows current load.


       apcupsd_loadgraph (height),(width) (gradient colour 1) (gradient colour
       2) (scale) (-t) (-l)
              History graph of current load.


       apcupsd_model

debian/changelog
conky (2.0.2-1) unstable; urgency=low

  * got temperature working again by talking to bme directly
  * added new variable ${battery_rate} to get charge/discharge rate

 -- Lance Colton <lance.colton@gmail.com>  Thu, 08 Mar 2012 00:45:00 -0700

conky (2.0.1-1) unstable; urgency=low

  * Added some power saving features, needs some testing to confirm this is working properly
    - Because conky is written in c (not c++) it can't directly access the system heartbeat timer so it uses another process to do this for it
      + Conky will look for another app I wrote and modified for this purpose, Proximus (>= v1.2.4)
      + If proximus-daemon is running conky will ask it to set up a heartbeat timer and proximus will send a dbus signal back for each heartbeat
      + update_heartbeat_min and update_heartbeat_max from the conf file will specify the approximate time between heartbeats
      + If the phone is discharging, Conky will skip heartbeats according to the value of update_heartbeat_battery_skip in the conf file
      + So, if update_heartbeat_battery_skip is 3, conky will update once every 4 heartbeats when discharging. default is 1 so set to 0 to disable.
      + If the screen turns off, update_skips_when_sleeping overrides update_heartbeat_battery_skip. default is 5!

 -- Lance Colton <lance.colton@gmail.com>  Fri, 27 Jan 2012 00:51:00 -0700

conky (2.0.0-1) unstable; urgency=low

  * Supports double buffering without dbe (an extension nokia removed from x11 in harmattan):
    - double buffers using pixmaps instead of dbe
      + spent a long time getting that working and fixing everything that broke when changing that
      + thanks to syrjala on meego.com forum for giving me a patch that got me started working on this
      + cairo / lua work, imlib2 works, and you need to install those 3 libraries also.
      + it doesn't actually crash when swiping to the background, it's nokia's idea that you meant to close it based on the physical orientation of the phone
      + .conf file needs some small fixes and i can't do anything about the i/o stats not working
      + i will add some more features soon

 -- Lance Colton <lance.colton@gmail.com>  Fri, 06 Jan 2012 21:00:00 -0700

conky (1.9.8-1) unstable; urgency=low

  * New maemo6 release:
    - initial release to support harmattan, don't install on maemo5! (sorry)
      + double buffering doesn't work, screen will flicker, this is not a conky bug. Nokia didn't include the double buffer extension module for x11, ask them!
      + temperature is not supported at the moment because bme doesn't report it anywhere, that was done with a 3rd party module on n900
      + cairo / lua work, imlib2 works, but these things aren't necessarily in the repositories so please google them to find them for now. you won't be able to install without those!
      + this app may crash when swiping it away, and may ignore the down swipe to close, i have no idea why. 
      + read/write io stats is a kernel feature that titan put into the power kernel on n900, feel free to bug nokia or replace the kernel with a 3rd party one to fix those from being 0
      + I spent lots of time fixing compiling issues and major issues in the lua and .conf files so didn't have time to optimize them for the 48 extra pixels of width or get everything perfect YET. Please be patient or contribute :)

 -- Lance Colton <lance.colton@gmail.com>  Tue, 08 Nov 2011 21:00:00 -0700

conky (1.9.7-1) unstable; urgency=low

  * New maemo release:
    - fixes for power kernel v47 
      + temperature and voltage are reported differently by this kernel and needed to be adjusted
      + temperature is now reported to the tenth of the degree and your conf file may need to be adjusted to account for this additional digit
      + enabled cairo / lua scripting

 -- Lance Colton <lance.colton@gmail.com>  Mon, 23 May 2011 16:00:00 -0700

conky (1.9.6-2) unstable; urgency=low

  * New maemo release:
    - too many changes to list
    - equivalent to 1.8.2 on master conky branch with lots of n900 fixes 
      + see http://talk.maemo.org/showthread.php?t=64434
      + the deb posted there is actually better as it supports lua, autobuilder can't work with that

 -- Lance Colton <lance.colton@gmail.com>  Thu, 21 Oct 2010 12:00:00 -0700
 
 conky (1.8.0-1) unstable; urgency=low

  * New upstream release:
    - add AF_UNIX socket support
    - fix sigsegv if config file is deleted (LP: #525926)
    - the following debian bugs are closed by this upload:
      + change automake1.10 to automake1.11 (Closes: #550929)
      + hwmon made compatible with kernels >= 2.6.31 (Closes: #556926)
      + text_buffer_size change is well documented (Closes: #519401)
      + fix diskio for not existing devices (Closes: #536557)
      + fix wrong mixer values on some systems (Closes: #540282)
      + fix minor memory leak (Closes: #566524)
      + fix some documentation error re. graphs (Closes: #564518)
      + add -p/--pause startup option (Closes: #513440)
      + fix documentation about mixer values (Closes: #538760)
  * This release is based on the Ubuntu package with the following changes
    necessary for Debian (Closes: #536320):
    - change control and rules to build correctly on non-Linux arches
      (Closes: #536326)
    - updated NEWS, descriptions in control and changelog
    - change archive area to contrib
  * Change priority of metapackage to extra
  * My utmost thanks go to Kapil Hari Paranjape for his packaging work and to
    Luca Falavigna for being so kind to review and sponsor this release

 -- Cesare Tirabassi <norsetto@ubuntu.com>  Sun, 28 Mar 2010 21:19:51 +0200

conky (1.8.0~git925e80ac-0ubuntu1) lucid; urgency=low

  * GIT snapshot release:
    - add argb (true transparency) support
    - add support for max_text_width
    - add IPv6 support to $tcp_portmon and mpd connections
    - add support for $pid variables ($pid_read, $pid_write, $pid_threads,

src/bme.c
/* 
 * File:   bme.C
 * Author: Lance
 * 
 * Created on March 6, 2012, 11:09 PM
 */

//#include "bme.h"

/*
** Copyright 2012, Alexey Roslyakov <alexey.roslyakov@newsycat.com>
**
** Licensed under the Apache License, Version 2.0 (the "License");
** you may not use this file except in compliance with the License.
** You may obtain a copy of the License at
**
**     http://www.apache.org/licenses/LICENSE-2.0
**
** Unless required by applicable law or agreed to in writing, software
** distributed under the License is distributed on an "AS IS" BASIS,
** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
** See the License for the specific language governing permissions and
** limitations under the License.
*/

// from here: https://gitorious.org/harmattan-goodies/bmecli/blobs/master/bmecli.c

#include <unistd.h>
#include <errno.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>

typedef unsigned int uint32;
typedef unsigned short uint16;
typedef unsigned char uint8;

//const char DefaultSocketPath[] = "/tmp/.bmesrv";

struct bme_reply { //22 items
  uint32     unknown1;
  uint32     unknown2;
  uint32     charger_type; // 0 - none, 2 - usbmax500, 3 - usbwall
  uint32     unknown4;
  uint32     unknown5;
  uint32     charging_time;
  uint32     unknown7;
  uint32     unknown8;
  uint32     battery_max_level; // - cur level?
  uint32     battery_cur_level; // /
  uint32     unknown11;
  uint32     unknown12;
  uint32     battery_max_capacity;
  uint32     battery_cur_capacity;
  uint32     battery_max_voltage;
  uint32     battery_cur_voltage;
  uint32     battery_pct_level;
  uint32     battery_temperature;
  int        battery_current;
  uint32     unknown20;
  uint32     battery_last_full_capacity;
  uint8      reserved[128];
};

static void dumpBuffer(const void *buf, size_t bufSize)
{
  int cnt = 0; char *p = (char*)buf;
  while (bufSize--)
    fprintf(stderr, "%02X%s", *p++, (++cnt & 15 ? " " : "\n") );
  fprintf(stderr, "\n");
}

const char* bmeClientGetData(const char *bmeSocketPath, struct bme_reply *bmeReply)
{
  //static int x = 0;
  //printf("%d\n", x); // outputs the value of x //always crashes on 23 or 24
  //x = x + 1;
    
  int s = socket(AF_LOCAL, SOCK_STREAM, 0);
  if (s <= 0)
    return "socket";
  struct sockaddr_un name;
  name.sun_family = AF_LOCAL;
  strcpy(name.sun_path, bmeSocketPath);
  if ( connect(s, (struct sockaddr *)&name, sizeof(name)) != 0)
    return "connect";

  char foo[256];
  size_t rd;
  // TODO: clean up this crap
  write(s, "SYNC\10\0\0\0BMentity", 16);
  rd = read(s, foo, sizeof(foo));
  //NORM_ERR("read %i bytes",rd); //reads 9 every time
  //fprintf(stderr, "rd=%zu\n", rd);
  write(s, "SYNC\4\0\0\0\3\200\0\0\n", 12);//crashes here
 
  rd = read(s, foo, 20);
  //NORM_ERR("bme socket read %i bytes",rd); //reads -1 right before it crashes
  //fprintf(stderr, "rd=%zu\n", rd);
  rd = read(s, bmeReply, sizeof(*bmeReply));
  //fprintf(stderr, "rd=%zu\n", rd);
 
  // FIXME: sometimes we get extra data in replies
  //NORM_ERR("read %i bytes",rd); //reads 0 right before it crashes
  if (rd > 128){
    memmove( bmeReply, ((char*) bmeReply) + rd - 128, 128);
  }
 // dumpBuffer(bmeReply, rd);
  close(s);
  return 0;
}

//int print_usage(const char *prog)
//{
//  fprintf(stderr, "Usage: %s <bme_socket_path>\n", prog);
//  fprintf(stderr, "sizeof=%zu\n", sizeof (struct bme_reply));
//
//  return -1;
//}

struct bme_reply getBattInfoFromBME()
{
//  if (argc != 2)
//    return print_usage(argv[0]);

  struct bme_reply bmeReply;
  
  memset(&bmeReply, 0, sizeof bmeReply);
  //NORM_ERR("bme 1");
  const char *failedOn = bmeClientGetData("/tmp/.bmesrv", &bmeReply);
  //NORM_ERR("bme 2");
  if (failedOn) {
    NORM_ERR("bme_client_get_data failed on %s: %s\n", failedOn, strerror(errno));
    //try again
    failedOn = 0;
    const char *failedOn = bmeClientGetData("/tmp/.bmesrv", &bmeReply);
    if (failedOn){//still failed
        NORM_ERR("failed 2x in a row @ bme");        
        bmeReply.battery_temperature = -1;
        bmeReply.battery_current = 0;
        return bmeReply;
    }
  }


/*
  fprintf(stderr, "BME stat\n");
  fprintf(stderr, "Unknown 1,2,4,5: %u %u %u %u\n", bmeReply.unknown1, bmeReply.unknown2, bmeReply.unknown4, bmeReply.unknown5);
  fprintf(stderr, "Charger type:\t\t%u\n", bmeReply.charger_type);
  fprintf(stderr, "Charging time:\t\t%u\n", bmeReply.charging_time);
  fprintf(stderr, "Battery temperature:\t%.2f\n", bmeReply.battery_temperature - 273.15f);
  fprintf(stderr, "Battery max. Level:\t%u\n", bmeReply.battery_max_level);
  fprintf(stderr, "Battery cur. Level:\t%u\n", bmeReply.battery_cur_level);
  fprintf(stderr, "Battery pct. Level:\t%u\n", bmeReply.battery_pct_level);
  fprintf(stderr, "Battery Max. Capacity:\t%u\n", bmeReply.battery_max_capacity);
  fprintf(stderr, "Battery Cur. Capacity:\t%u\n", bmeReply.battery_cur_capacity);
  fprintf(stderr, "Battery Last Full Cap:\t%u\n", bmeReply.battery_last_full_capacity);
  fprintf(stderr, "Battery Max. Voltage:\t%u\n", bmeReply.battery_max_voltage);
  fprintf(stderr, "Battery Cur. Voltage:\t%u\n", bmeReply.battery_cur_voltage);
  fprintf(stderr, "Battery Current:\t%d\n", bmeReply.battery_current);
  fprintf(stderr, "Unknown 7,8:\t%u %u\n", bmeReply.unknown7, bmeReply.unknown7);
  fprintf(stderr, "Unknown 11,12:\t%u %u\n", bmeReply.unknown11, bmeReply.unknown12);
  fprintf(stderr, "Unknown20 (timestamp?):\t%u\n", bmeReply.unknown20);
  fprintf(stderr, "\n");  
*/


  return bmeReply;
}


